// @redhat-cloud-services/frontend-components/ConditionalFilter/ConditionalFilter.spec.ct.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("@cypress/react");
var ConditionalFilter_1 = __importDefault(require("./ConditionalFilter"));
describe('ConditionalFilter component', function () {
    var config = [
        {
            label: 'text',
        },
        {
            label: 'checkbox',
            type: 'checkbox',
            filterValues: {
                items: [
                    {
                        label: 'cfilter1',
                    },
                    {
                        label: 'cfilter2',
                    },
                ],
            },
        },
        {
            label: 'radio',
            type: 'radio',
            filterValues: {
                items: [
                    {
                        label: 'rfilter1',
                    },
                    {
                        label: 'rfilter2',
                    },
                ],
            },
        },
        {
            label: 'group',
            type: 'group',
            filterValues: {
                groups: [
                    {
                        label: 'group1',
                        type: 'checkbox',
                        groupSelectable: true,
                        items: [
                            {
                                label: 'gcitem1',
                                type: 'checkbox',
                            },
                            {
                                label: 'gcitem2',
                                type: 'checkbox',
                            },
                        ],
                    },
                ],
            },
        },
    ];
    it('renders empty', function () {
        (0, react_1.mount)((0, jsx_runtime_1.jsx)(ConditionalFilter_1.default, {}));
        cy.get('.ins-c-conditional-filter');
    });
    it('renders disabled', function () {
        (0, react_1.mount)((0, jsx_runtime_1.jsx)(ConditionalFilter_1.default, { isDisabled: true, items: config }));
        cy.get('.pf-c-dropdown').click({ force: true });
        cy.get('.pf-c-dropdown__menu').should('not.exist');
    });
    it('renders with data', function () {
        (0, react_1.mount)((0, jsx_runtime_1.jsx)(ConditionalFilter_1.default, { items: config }));
        cy.get('.ins-c-conditional-filter');
        cy.get('.pf-c-dropdown').click();
        cy.get('.pf-c-dropdown__menu').children().should('have.length', 4);
    });
    it('filter changes on click', function () {
        (0, react_1.mount)((0, jsx_runtime_1.jsx)(ConditionalFilter_1.default, { items: config }));
        cy.get('.pf-c-dropdown').click();
        cy.get('.pf-c-dropdown__menu').children().eq(1).click();
        cy.get('.pf-c-dropdown').should('not.contain.text', 'Text');
        cy.get('.pf-c-select').click();
        cy.get('.pf-c-select__menu');
    });
    it('onChange called', function () {
        var ocSpy = cy.spy().as('ocSpy');
        (0, react_1.mount)((0, jsx_runtime_1.jsx)(ConditionalFilter_1.default, { items: config, onChange: ocSpy }));
        cy.get('.pf-c-dropdown').click();
        cy.get('.pf-c-dropdown__menu').children().eq(2).click();
        cy.get('@ocSpy').should('have.been.called');
    });
});
