// @redhat-cloud-services/frontend-components/ConditionalFilter/GroupFilter.spec.ct.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("@cypress/react");
var GroupFilter_1 = __importDefault(require("./GroupFilter"));
describe('GroupFilter component', function () {
    var config = {
        groups: [
            {
                label: 'Group1',
                items: [
                    {
                        label: 'Item11',
                    },
                    {
                        label: 'Item12',
                    },
                ],
            },
            {
                label: 'Group2',
                type: 'radio',
                items: [
                    {
                        label: 'Item21',
                    },
                    {
                        label: 'Item22',
                    },
                ],
            },
            {
                label: 'Group3',
                type: 'checkbox',
                items: [
                    {
                        label: 'Item31',
                    },
                    {
                        label: 'Item32',
                    },
                ],
            },
            {
                label: 'Group4',
                groupSelectable: true,
                items: [
                    {
                        label: 'Item41',
                    },
                    {
                        label: 'Item42',
                    },
                ],
            },
        ],
        items: [
            {
                label: 'Group5',
                type: 'treeView',
                children: [
                    {
                        label: 'Item51',
                    },
                    {
                        label: 'Item52',
                    },
                ],
            },
        ],
    };
    it('renders empty', function () {
        (0, react_1.mount)((0, jsx_runtime_1.jsx)(GroupFilter_1.default, {}));
        cy.get('.pf-c-menu-toggle');
    });
    it('renders with placeholder', function () {
        (0, react_1.mount)((0, jsx_runtime_1.jsx)(GroupFilter_1.default, { placeholder: "foo" }));
        cy.get('.pf-c-menu-toggle__text').should('contain', 'foo');
    });
    it('renders with data', function () {
        (0, react_1.mount)((0, jsx_runtime_1.jsx)(GroupFilter_1.default, __assign({}, config, { showMoreTitle: "hello", onShowMore: function () { return undefined; } })));
        cy.get('.pf-c-menu-toggle__controls').click();
        cy.get('.pf-c-menu__item').should('have.length', 11);
    });
    it('onChange called', function () {
        var ocSpy = cy.spy().as('ocSpy');
        (0, react_1.mount)((0, jsx_runtime_1.jsx)(GroupFilter_1.default, __assign({}, config, { onChange: ocSpy })));
        cy.get('.pf-c-menu-toggle__controls').click();
        cy.get('.pf-c-menu__item').eq(5).click();
        cy.get('@ocSpy').should('have.been.called');
    });
});
