// @redhat-cloud-services/frontend-components-utilities/RBAC/RBAC.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RBACContext = exports.initialPermissions = exports.hasAllPermissions = exports.doesHavePermissions = exports.getRBAC = void 0;
var react_1 = require("react");
function getRBAC(applicationName, disableCache) {
    var _a, _b, _c, _d, _e;
    if (applicationName === void 0) { applicationName = ''; }
    if (disableCache === void 0) { disableCache = false; }
    return __awaiter(this, void 0, void 0, function () {
        var insights, user;
        var _f;
        return __generator(this, function (_g) {
            switch (_g.label) {
                case 0:
                    insights = window.insights;
                    return [4, ((_b = (_a = insights === null || insights === void 0 ? void 0 : insights.chrome) === null || _a === void 0 ? void 0 : _a.auth) === null || _b === void 0 ? void 0 : _b.getUser())];
                case 1:
                    user = _g.sent();
                    _f = {
                        isOrgAdmin: ((_d = (_c = user === null || user === void 0 ? void 0 : user.identity) === null || _c === void 0 ? void 0 : _c.user) === null || _d === void 0 ? void 0 : _d.is_org_admin) || false
                    };
                    return [4, ((_e = insights === null || insights === void 0 ? void 0 : insights.chrome) === null || _e === void 0 ? void 0 : _e.getUserPermissions(applicationName, disableCache))];
                case 2: return [2, (_f.permissions = (_g.sent()) || null,
                        _f)];
            }
        });
    });
}
exports.getRBAC = getRBAC;
function isAccessType(permission) {
    return typeof permission === 'object';
}
function doesHavePermissions(userPermissions, permissionList) {
    if (!userPermissions) {
        return false;
    }
    return userPermissions.some(function (access) {
        return permissionList.includes(isAccessType(access) ? access === null || access === void 0 ? void 0 : access.permission : access);
    });
}
exports.doesHavePermissions = doesHavePermissions;
function hasAllPermissions(userPermissions, permissionList) {
    if (!userPermissions) {
        return false;
    }
    return permissionList.every(function (permission) {
        return userPermissions.some(function (access) {
            var _a;
            var accessArray = ((_a = (isAccessType(access) ? access === null || access === void 0 ? void 0 : access.permission : access)) === null || _a === void 0 ? void 0 : _a.split(':')) || [];
            var permissionArray = permission.split(':');
            var hasAccess = accessArray.slice(0).reduce(function (acc, curr, index, array) {
                if (acc === false) {
                    array.splice(index);
                    return acc;
                }
                return curr === '*' || curr === (permissionArray === null || permissionArray === void 0 ? void 0 : permissionArray[index]);
            }, true);
            return hasAccess || accessArray.join(':') === permission;
        });
    });
}
exports.hasAllPermissions = hasAllPermissions;
exports.initialPermissions = {
    isLoading: true,
    isOrgAdmin: false,
    permissions: [],
};
exports.RBACContext = (0, react_1.createContext)(exports.initialPermissions);
exports.default = getRBAC;
