// @redhat-cloud-services/frontend-components-utilities/CypressUtils/PaginationUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SORTING_ORDERS = exports.PAGINATION_VALUES = exports.DEFAULT_ROW_COUNT = exports.changePagination = exports.checkPaginationValues = exports.checkPaginationTotal = exports.itemsPerPage = void 0;
var selectors_1 = require("./selectors");
var DEFAULT_ROW_COUNT = 20;
exports.DEFAULT_ROW_COUNT = DEFAULT_ROW_COUNT;
var PAGINATION_VALUES = [10, 20, 50, 100];
exports.PAGINATION_VALUES = PAGINATION_VALUES;
var SORTING_ORDERS = ['ascending', 'descending'];
exports.SORTING_ORDERS = SORTING_ORDERS;
function itemsPerPage(totalLength, pageSize) {
    if (pageSize === void 0) { pageSize = DEFAULT_ROW_COUNT; }
    var items = totalLength;
    var array = [];
    while (items > 0) {
        var remain = items - pageSize;
        var v = remain > 0 ? pageSize : items;
        array.push(v);
        items = remain;
    }
    return array;
}
exports.itemsPerPage = itemsPerPage;
function checkPaginationTotal(n) {
    return cy.get('.pf-c-options-menu__toggle-text').find('b').eq(1).should('have.text', n);
}
exports.checkPaginationTotal = checkPaginationTotal;
function checkPaginationValues(expectedValues) {
    cy.get(selectors_1.TOOLBAR).find(selectors_1.PAGINATION_MENU).find(selectors_1.DROPDOWN_TOGGLE).click();
    cy.get(selectors_1.TOOLBAR)
        .find(selectors_1.PAGINATION_MENU)
        .find('ul[class=pf-c-options-menu__menu]')
        .find('li')
        .each(function ($el, index) {
        cy.wrap($el).should('have.text', "".concat(expectedValues[index], " per page"));
    });
}
exports.checkPaginationValues = checkPaginationValues;
function changePagination(paginationValue) {
    cy.get(selectors_1.TOOLBAR).find(selectors_1.PAGINATION_MENU).find(selectors_1.DROPDOWN_TOGGLE).click();
    return cy.get(selectors_1.TOOLBAR).find(selectors_1.PAGINATION_MENU).find('ul[class=pf-c-options-menu__menu]').find(selectors_1.DROPDOWN_ITEM).contains("".concat(paginationValue)).click();
}
exports.changePagination = changePagination;
