// @redhat-cloud-services/frontend-components-utilities/CypressUtils/selectors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FILTER_TOGGLE = exports.FILTERS_DROPDOWN = exports.TITLE = exports.ROWS_TOGGLER = exports.TABLE_HEADER = exports.TABLE = exports.TOOLBAR_FILTER = exports.TBODY = exports.DROPDOWN_ITEM = exports.DROPDOWN_TOGGLE = exports.TEXT_INPUT = exports.CHECKBOX = exports.MODAL = exports.DROPDOWN = exports.PAGINATION_MENU = exports.PAGINATION = exports.ROW = exports.CHIP = exports.CHIP_GROUP = exports.TOOLBAR = exports.ouiaId = void 0;
var TOOLBAR = 'div[id="ins-primary-data-toolbar"]';
exports.TOOLBAR = TOOLBAR;
var CHIP_GROUP = 'div[data-ouia-component-type="PF4/ChipGroup"]';
exports.CHIP_GROUP = CHIP_GROUP;
var CHIP = '[data-ouia-component-type="PF4/Chip"]';
exports.CHIP = CHIP;
var ROW = '[data-ouia-component-type="PF4/TableRow"]:not([class~="pf-c-table__expandable-row"])';
exports.ROW = ROW;
var PAGINATION = 'div[data-ouia-component-type="PF4/Pagination"]';
exports.PAGINATION = PAGINATION;
var PAGINATION_MENU = 'div[data-ouia-component-type="PF4/PaginationOptionsMenu"]';
exports.PAGINATION_MENU = PAGINATION_MENU;
var DROPDOWN = '[data-ouia-component-type="PF4/Dropdown"]';
exports.DROPDOWN = DROPDOWN;
var MODAL = '[data-ouia-component-type="PF4/ModalContent"]';
exports.MODAL = MODAL;
var CHECKBOX = '[data-ouia-component-type="PF4/Checkbox"]';
exports.CHECKBOX = CHECKBOX;
var TEXT_INPUT = '[data-ouia-component-type="PF4/TextInput"]';
exports.TEXT_INPUT = TEXT_INPUT;
var DROPDOWN_TOGGLE = '[data-ouia-component-type="PF4/DropdownToggle"]';
exports.DROPDOWN_TOGGLE = DROPDOWN_TOGGLE;
var DROPDOWN_ITEM = '[data-ouia-component-type="PF4/DropdownItem"]';
exports.DROPDOWN_ITEM = DROPDOWN_ITEM;
var TBODY = 'tbody[role=rowgroup]';
exports.TBODY = TBODY;
var TOOLBAR_FILTER = '.ins-c-primary-toolbar__filter';
exports.TOOLBAR_FILTER = TOOLBAR_FILTER;
var TABLE = 'table';
exports.TABLE = TABLE;
var TABLE_HEADER = 'thead';
exports.TABLE_HEADER = TABLE_HEADER;
var ROWS_TOGGLER = "".concat(TABLE_HEADER, " .pf-c-table__toggle");
exports.ROWS_TOGGLER = ROWS_TOGGLER;
var TITLE = '[data-ouia-component-type="PF4/Title"]';
exports.TITLE = TITLE;
var ouiaId = function (id) { return "[data-ouia-component-id=\"".concat(id, "\"]"); };
exports.ouiaId = ouiaId;
var FILTERS_DROPDOWN = 'ul[class=pf-c-dropdown__menu]';
exports.FILTERS_DROPDOWN = FILTERS_DROPDOWN;
var FILTER_TOGGLE = 'button[class=pf-c-select__toggle]';
exports.FILTER_TOGGLE = FILTER_TOGGLE;
