// @redhat-cloud-services/frontend-components-notifications/redux/actions/notifications.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearNotifications = exports.removeNotification = exports.addNotification = void 0;
var action_types_1 = require("./action-types");
function generateID(type) {
    return 'cloud-services' + type + '-' + new Date().getTime() + Math.random().toString(36).slice(2);
}
var addNotification = function (notification) { return ({
    type: action_types_1.ADD_NOTIFICATION,
    payload: __assign({ id: generateID('notification') }, notification),
}); };
exports.addNotification = addNotification;
var removeNotification = function (index) { return ({
    type: action_types_1.REMOVE_NOTIFICATION,
    payload: index,
}); };
exports.removeNotification = removeNotification;
var clearNotifications = function () { return ({
    type: action_types_1.CLEAR_NOTIFICATIONS,
}); };
exports.clearNotifications = clearNotifications;
exports.default = {
    addNotification: exports.addNotification,
    removeNotification: exports.removeNotification,
    clearNotifications: exports.clearNotifications,
};
