// @redhat-cloud-services/frontend-components-utilities/esm/CypressUtils/PaginationUtils.js
import { DROPDOWN_ITEM, DROPDOWN_TOGGLE, PAGINATION_MENU, TOOLBAR } from './selectors';
var DEFAULT_ROW_COUNT = 20;
var PAGINATION_VALUES = [10, 20, 50, 100];
var SORTING_ORDERS = ['ascending', 'descending'];
export function itemsPerPage(totalLength, pageSize) {
    if (pageSize === void 0) { pageSize = DEFAULT_ROW_COUNT; }
    var items = totalLength;
    var array = [];
    while (items > 0) {
        var remain = items - pageSize;
        var v = remain > 0 ? pageSize : items;
        array.push(v);
        items = remain;
    }
    return array;
}
export function checkPaginationTotal(n) {
    return cy.get('.pf-c-options-menu__toggle-text').find('b').eq(1).should('have.text', n);
}
export function checkPaginationValues(expectedValues) {
    cy.get(TOOLBAR).find(PAGINATION_MENU).find(DROPDOWN_TOGGLE).click();
    cy.get(TOOLBAR)
        .find(PAGINATION_MENU)
        .find('ul[class=pf-c-options-menu__menu]')
        .find('li')
        .each(function ($el, index) {
        cy.wrap($el).should('have.text', "".concat(expectedValues[index], " per page"));
    });
}
export function changePagination(paginationValue) {
    cy.get(TOOLBAR).find(PAGINATION_MENU).find(DROPDOWN_TOGGLE).click();
    return cy.get(TOOLBAR).find(PAGINATION_MENU).find('ul[class=pf-c-options-menu__menu]').find(DROPDOWN_ITEM).contains("".concat(paginationValue)).click();
}
export { DEFAULT_ROW_COUNT, PAGINATION_VALUES, SORTING_ORDERS };
