// @redhat-cloud-services/frontend-components-utilities/esm/CypressUtils/selectors.js
var TOOLBAR = 'div[id="ins-primary-data-toolbar"]';
var CHIP_GROUP = 'div[data-ouia-component-type="PF4/ChipGroup"]';
var CHIP = '[data-ouia-component-type="PF4/Chip"]';
var ROW = '[data-ouia-component-type="PF4/TableRow"]:not([class~="pf-c-table__expandable-row"])';
var PAGINATION = 'div[data-ouia-component-type="PF4/Pagination"]';
var PAGINATION_MENU = 'div[data-ouia-component-type="PF4/PaginationOptionsMenu"]';
var DROPDOWN = '[data-ouia-component-type="PF4/Dropdown"]';
var MODAL = '[data-ouia-component-type="PF4/ModalContent"]';
var CHECKBOX = '[data-ouia-component-type="PF4/Checkbox"]';
var TEXT_INPUT = '[data-ouia-component-type="PF4/TextInput"]';
var DROPDOWN_TOGGLE = '[data-ouia-component-type="PF4/DropdownToggle"]';
var DROPDOWN_ITEM = '[data-ouia-component-type="PF4/DropdownItem"]';
var TBODY = 'tbody[role=rowgroup]';
var TOOLBAR_FILTER = '.ins-c-primary-toolbar__filter';
var TABLE = 'table';
var TABLE_HEADER = 'thead';
var ROWS_TOGGLER = "".concat(TABLE_HEADER, " .pf-c-table__toggle");
var TITLE = '[data-ouia-component-type="PF4/Title"]';
var ouiaId = function (id) { return "[data-ouia-component-id=\"".concat(id, "\"]"); };
var FILTERS_DROPDOWN = 'ul[class=pf-c-dropdown__menu]';
var FILTER_TOGGLE = 'button[class=pf-c-select__toggle]';
export { ouiaId, TOOLBAR, CHIP_GROUP, CHIP, ROW, PAGINATION, PAGINATION_MENU, DROPDOWN, MODAL, CHECKBOX, TEXT_INPUT, DROPDOWN_TOGGLE, DROPDOWN_ITEM, TBODY, TOOLBAR_FILTER, TABLE, TABLE_HEADER, ROWS_TOGGLER, TITLE, FILTERS_DROPDOWN, FILTER_TOGGLE, };
