// @redhat-cloud-services/frontend-components/Battery/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NullBattery = exports.LowBattery = exports.MediumBattery = exports.HighBattery = exports.CriticalBattery = exports.Battery = exports.default = void 0;
var Battery_1 = require("./Battery");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return __importDefault(Battery_1).default; } });
var Battery_2 = require("./Battery");
Object.defineProperty(exports, "Battery", { enumerable: true, get: function () { return __importDefault(Battery_2).default; } });
var CriticalBattery_1 = require("./CriticalBattery");
Object.defineProperty(exports, "CriticalBattery", { enumerable: true, get: function () { return __importDefault(CriticalBattery_1).default; } });
var HighBattery_1 = require("./HighBattery");
Object.defineProperty(exports, "HighBattery", { enumerable: true, get: function () { return __importDefault(HighBattery_1).default; } });
var MediumBattery_1 = require("./MediumBattery");
Object.defineProperty(exports, "MediumBattery", { enumerable: true, get: function () { return __importDefault(MediumBattery_1).default; } });
var LowBattery_1 = require("./LowBattery");
Object.defineProperty(exports, "LowBattery", { enumerable: true, get: function () { return __importDefault(LowBattery_1).default; } });
var NullBattery_1 = require("./NullBattery");
Object.defineProperty(exports, "NullBattery", { enumerable: true, get: function () { return __importDefault(NullBattery_1).default; } });
