// @redhat-cloud-services/frontend-components-utilities/RouterParams/RouterParams.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = __importDefault(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var react_router_dom_1 = require("react-router-dom");
var react_redux_1 = require("react-redux");
var isEqual_1 = __importDefault(require("lodash/isEqual"));
var RouterParams = (function (_super) {
    __extends(RouterParams, _super);
    function RouterParams() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RouterParams.prototype.componentDidMount = function () {
        var _a = this.props, _b = _a.match, params = _b.params, path = _b.path, url = _b.url, onPathChange = _a.onPathChange, location = _a.location;
        if ((0, react_router_dom_1.matchPath)(location.pathname, { path: url, exact: true })) {
            onPathChange &&
                onPathChange({
                    params: params,
                    path: path,
                });
        }
    };
    RouterParams.prototype.componentDidUpdate = function () {
        var _a = this.props, _b = _a.match, params = _b.params, path = _b.path, url = _b.url, onPathChange = _a.onPathChange, location = _a.location, routerData = _a.routerData;
        if (routerData && (!(0, isEqual_1.default)(params, routerData.params) || path !== routerData.path)) {
            if ((0, react_router_dom_1.matchPath)(location.pathname, { path: url, exact: true })) {
                onPathChange &&
                    onPathChange({
                        params: params,
                        path: path,
                    });
            }
        }
    };
    RouterParams.prototype.render = function () {
        var _a = this.props, onPathChange = _a.onPathChange, routerData = _a.routerData, staticContext = _a.staticContext, Component = _a.Component, props = __rest(_a, ["onPathChange", "routerData", "staticContext", "Component"]);
        return (0, jsx_runtime_1.jsx)(Component, __assign({}, props));
    };
    return RouterParams;
}(react_1.default.Component));
RouterParams.propTypes = {
    match: prop_types_1.default.shape({ params: prop_types_1.default.object, path: prop_types_1.default.string, url: prop_types_1.default.string }).isRequired,
    onPathChange: prop_types_1.default.func,
    location: prop_types_1.default.shape({ pathname: prop_types_1.default.string.isRequired }).isRequired,
    routerData: prop_types_1.default.shape({
        params: prop_types_1.default.object,
        path: prop_types_1.default.string,
    }).isRequired,
    staticContext: prop_types_1.default.any,
    Component: prop_types_1.default.any,
};
var routerParams = function (Component) {
    return (0, react_router_dom_1.withRouter)((0, react_redux_1.connect)(function (_a) {
        var routerData = _a.routerData;
        return ({ routerData: routerData });
    }, function (dispatch) { return ({
        onPathChange: function (data) {
            return dispatch({
                type: '@@INSIGHTS-CORE/NAVIGATION',
                payload: data,
            });
        },
    }); })(function (props) { return (0, jsx_runtime_1.jsx)(RouterParams, __assign({}, props, { Component: Component })); }));
};
exports.default = routerParams;
